File: lib/core/utils/contact_email.dart
/// Política única para o e-mail público de contato do perfil.
///
/// O e-mail usado para autenticação pode ser um alias privado da Apple ou
/// um endereço sintético do login por telefone. Esses endereços continuam
/// válidos para login, mas não devem ser publicados em currículos nem usados
/// como contato por recrutadores.
class ContactEmail {
  ContactEmail._();

  static const Set<String> _applePrivateRelayDomains = {
    'privaterelay.appleid.com',
    'private.icloud.com',
  };

  static final RegExp _validEmail = RegExp(r'^[^@\s]+@[^@\s]+\.[^@\s]+$');

  static String normalize(String? value) => value?.trim().toLowerCase() ?? '';

  static String _domain(String? value) {
    final normalized = normalize(value);
    final separator = normalized.lastIndexOf('@');
    return separator < 0 ? '' : normalized.substring(separator + 1);
  }

  static bool isValid(String? value) {
    final normalized = normalize(value);
    return normalized.isNotEmpty && _validEmail.hasMatch(normalized);
  }

  static bool isApplePrivateRelay(String? value) =>
      _applePrivateRelayDomains.contains(_domain(value));

  static bool isSyntheticAuthEmail(String? value) {
    final normalized = normalize(value);
    final separator = normalized.indexOf('@');
    if (separator <= 0 || _domain(normalized) != 'stage.app') return false;
    final localPart = normalized.substring(0, separator);
    return localPart.startsWith('phone_') && localPart.length > 'phone_'.length;
  }

  static bool isPrivateOrSynthetic(String? value) =>
      isApplePrivateRelay(value) || isSyntheticAuthEmail(value);

  /// Um e-mail publicável precisa ter formato válido e não pode ser um
  /// identificador privado/sintético de autenticação.
  static bool isUsable(String? value) =>
      isValid(value) && !isPrivateOrSynthetic(value);

  /// Defesa de última linha para saídas públicas, como currículos.
  static String resumeValueOrEmpty(String? value) =>
      isUsable(value) ? normalize(value) : '';

  /// Resolve o valor inicial sem misturar identidade de login com contato.
  /// Um valor inseguro é ignorado e a próxima fonte é considerada.
  static String resolveInitial({
    String? profileEmail,
    String? extractedEmail,
    String? authEmail,
  }) {
    for (final candidate in [profileEmail, extractedEmail, authEmail]) {
      if (isUsable(candidate)) return normalize(candidate);
    }
    return '';
  }
}

File: supabase/functions/_shared/contact_email.ts
/**
 * Política compartilhada para o e-mail público do perfil.
 *
 * Endereços privados/sintéticos continuam válidos como identidade de login,
 * mas não podem aparecer em currículos ou superfícies de recrutamento.
 */

const APPLE_PRIVATE_RELAY_DOMAINS = new Set([
  'privaterelay.appleid.com',
  'private.icloud.com',
]);

const VALID_EMAIL = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;

export function normalizeContactEmail(value: unknown): string {
  return typeof value === 'string' ? value.trim().toLowerCase() : '';
}

function emailDomain(value: unknown): string {
  const normalized = normalizeContactEmail(value);
  const separator = normalized.lastIndexOf('@');
  return separator < 0 ? '' : normalized.slice(separator + 1);
}

export function isValidContactEmail(value: unknown): boolean {
  const normalized = normalizeContactEmail(value);
  return normalized.length > 0 && VALID_EMAIL.test(normalized);
}

export function isApplePrivateRelayEmail(value: unknown): boolean {
  return APPLE_PRIVATE_RELAY_DOMAINS.has(emailDomain(value));
}

export function isSyntheticAuthEmail(value: unknown): boolean {
  const normalized = normalizeContactEmail(value);
  const separator = normalized.indexOf('@');
  if (separator <= 0 || emailDomain(normalized) !== 'stage.app') return false;
  const localPart = normalized.slice(0, separator);
  return localPart.startsWith('phone_') && localPart.length > 'phone_'.length;
}

export function isUsableContactEmail(value: unknown): boolean {
  return isValidContactEmail(value) &&
    !isApplePrivateRelayEmail(value) &&
    !isSyntheticAuthEmail(value);
}

export function publicContactEmailOrEmpty(value: unknown): string {
  return isUsableContactEmail(value) ? normalizeContactEmail(value) : '';
}

/**
 * Importação preserva sempre um contato público escolhido pelo usuário.
 * Um valor extraído só preenche perfil vazio/privado e nunca introduz relay.
 * `null` faz o RPC manter o valor existente por sua semântica de COALESCE.
 */
export function resolveImportedContactEmail(
  existingValue: unknown,
  importedValue: unknown,
): string | null {
  const existing = publicContactEmailOrEmpty(existingValue);
  if (existing) return existing;

  const imported = publicContactEmailOrEmpty(importedValue);
  return imported || null;
}
